### PR TITLE
server: enable replication setting when only have one default placeme…

### PR DESCRIPTION
**Pull request by duduainankai**
…nt rule

Signed-off-by: Zheming Li <nkdudu@126.com>

### What problem does this PR solve?

#3140

### What is changed and how it works?

- When update max replica and placement rule enabled, check if have only one default rule.

### Check List

Tests

- Unit test

### Release note

- No release note

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -857,6 +858,11 @@ func (s *Server) SetReplicationConfig(cfg config.ReplicationConfig) error {
 	return nil
 }
 
+func (s *Server) isOnlyDefaultRule() bool {
+	rules := s.GetRaftCluster().GetRuleManager().GetAllRules()
+	return len(rules) == 1 && rules[0].GroupID == "pd" && rules[0].ID == "default"
```
Comment (rleungx) [inline server/server.go:863]: What if we overwrite this default rule?

File: tests/pdctl/config/config_test.go [hunk 2]
```diff
@@ -567,3 +563,78 @@ func (s *configTestSuite) TestReplicationMode(c *C) {
...
+	c.Assert(err, IsNil)
+	ioutil.WriteFile(fname, b, 0644)
+	_, _, err = pdctl.ExecuteCommandC(cmd, "-u", pdAddr, "config", "placement-rules", "save", "--in="+fname)
+	c.Assert(err, IsNil)
+
+	_, output, err = pdctl.ExecuteCommandC(cmd, "-u", pdAddr, "config", "set", "max-replicas", "4")
+	c.Assert(err, IsNil)
+	c.Assert(strings.Contains(string(output), "please update rule instead"), IsTrue)
+	checkMaxReplicas(3)
+
```
Comment (rleungx) [inline tests/pdctl/config/config_test.go:639]: Please remove the empty line.

File: server/server.go (hunk 1 above)
Comment (duduainankai) [reply server/server.go:863]: So what's the definition of 'default rule' ? 

I thought it was only defined by GroupID and ID.

**Comment by disksing** [comment]
em, I think you need not only allow setting the config but also need to sync the new configuration with the default rule.

**Comment by Yisaer** [comment]
I think the origin issue could be described like: 
```
>> pd-ctl config set max-replicas <group-id> <rule-id> 1
successful!
```

so that the pd-ctl could edit max-relicas for each placement rule. WDYT @nolouch

**Comment by nolouch** [comment]
@Yisaer This is an extension, but I hope it can be compatible with the original command.

**Comment by nolouch** [comment]
Hi, @duduainankai , ci failed.
```
FAIL: cluster_test.go:47: testClusterSuite.TestCluster

cluster_test.go:63:
    // Cannot set the replicas when placement rules is enabled.
    c.Assert(s.svr.SetReplicationConfig(r), NotNil)
... value = nil
```

File: tests/pdctl/config/config_test.go [hunk 3]
```diff
@@ -567,3 +563,78 @@ func (s *configTestSuite) TestReplicationMode(c *C) {
...
+			GroupID: "pd",
+			ID:      "test1",
+			Role:    "voter",
+			Count:   1,
+		},
+	}
+	b, err := json.Marshal(rules)
+	c.Assert(err, IsNil)
+	ioutil.WriteFile(fname, b, 0644)
+	_, _, err = pdctl.ExecuteCommandC(cmd, "-u", pdAddr, "config", "placement-rules", "save", "--in="+fname)
```
Comment (ZenoTan) [inline tests/pdctl/config/config_test.go:632]: Also check replicas here?

**Comment by duduainankai** [comment]
@nolouch @ZenoTan Got

But I still try to understand the comments:

> What if we overwrite this default rule?

> em, I think you need not only allow setting the config but also need to sync the new configuration with the default rule.

and a little busy these days, maybe need more time, sorry about that :( .

File: server/api/cluster_test.go [hunk 4]
```diff
@@ -56,12 +56,8 @@ func (s *testClusterSuite) TestCluster(c *C) {
 
 	c2 := &metapb.Cluster{}
 	r := config.ReplicationConfig{
-		EnablePlacementRules: true,
```
Comment (rleungx) [inline server/api/cluster_test.go:59]: Why remove it?

Comment (duduainankai) [reply server/api/cluster_test.go:59]: This is the "only one default rule when enable placement rules", so `c.Assert(s.svr.SetReplicationConfig(r), NotNil)` is no longer correct here.

And the similar test is already done in `config_test`, so I think just remove the code to simply fix the test is ok.

**Review by rleungx — APPROVED** [review]
Overall, LGTM, but there is another case: once we change the default rule, and then we disable the placement rule, the original max replicas will not match.

**Review by nolouch — APPROVED** [review]
LGTM. thanks

**Pull request merged**